File: engine.py
from typing import List, Tuple, Optional
import math
from board import Board
from pieces import Piece, General
from supports import EvaluateSet, Vector, Color
import random

class Engine:
    def __init__(self, eval_set:EvaluateSet, depth: int = 3, debug: bool = False):
        self.debug = debug
        self.depth = depth
        self.eval_set = eval_set

    def get_best_move(self, board: Board, current_player: Color, is_random:bool = False) -> Optional[Tuple[Vector, Vector]]:
        alpha = -math.inf
        beta = math.inf
        if is_random:
            best_moves = []
        pieces = board.get_reds() if current_player == Color.RED else board.get_blacks()
        moves = []
        for piece in pieces:
            from_pos = piece.get_position()
            valid_moves = board.get_piece_valid_moves(piece, check=True)
            
            for to_pos in valid_moves:
                target_square = board.get_square(to_pos)
                is_capture = target_square.get_piece() is not None
                
                quick_eval = board.ghost_test(from_pos, to_pos, 
                                            lambda b: b.evaluate(self.eval_set))
                eval_bonus = 0
                if is_capture:
                    captured_piece = target_square.get_piece()
                    eval_bonus += captured_piece.get_value() * 0.5
                
                if isinstance(piece, General):
                    eval_bonus += 0.3
                
                moves.append((quick_eval + eval_bonus, from_pos, to_pos))
        moves.sort(reverse=(current_player == Color.RED), key=lambda x: x[0])

        if len(moves):
            best_move = (moves[0][1], moves[0][2])
        else:
            best_move = None

        for move in moves:
            bonus, from_pos, to_pos = move

            if self.debug:
                print()
                print("=============================================================================================================================")
                print(f"ENGINE for {current_player.name} with {from_pos}->{to_pos}:")

            move_value = self.minimax(
                    board=board,
                    from_pos=from_pos,
                    to_pos=to_pos,
                    depth=self.depth,
                    alpha=alpha,
                    beta=beta,
                    maximizing_player=(current_player == Color.RED))

            if self.debug:
                print(f"ENGINE EVALUATION for {current_player.name} after {from_pos}->{to_pos}: {current_player.opposite().name} can get {move_value}")
                print("=============================================================================================================================")
                print()

            if current_player == Color.RED and move_value > alpha:
                if is_random and ((move_value / alpha) if alpha else 2) < 1.05:
                    best_moves.append((from_pos, to_pos))
                else:
                    best_move = (from_pos, to_pos)
                alpha = move_value
                
            elif current_player == Color.BLACK and beta > move_value:
                if is_random and ((beta / move_value) if move_value else 2) < 1.05:
                    best_moves.append((from_pos, to_pos))
                else:
                    best_move = (from_pos, to_pos)
                beta = move_value
                
                if abs(move_value) == math.inf:
                    print("FOUND CHECKMATE")
                    break
    
        move_value = alpha if current_player==Color.RED else beta
        if is_random:
            result = (move_value,
                    self.handle_gameover(board, current_player) if best_moves==[] else random.choice(best_moves))
        else:
            result = (move_value,
                    best_move if best_move else self.handle_gameover(board, current_player))

        return result

    def handle_gameover(self, board: Board, color: Color):
        board.debug = True
        team = board.get_reds() if color == Color.RED else board.get_blacks()
        for piece in team:
            vds = board.get_piece_valid_moves(piece, check=True)
            print(f"ENGINE GAMEOVER: {piece} valid moves = {vds}")
        board.debug = self.debug
        return None

    def minimax(self, 
               board: Board,
               from_pos: Vector,
               to_pos: Vector,
               depth: int,
               alpha: float,
               beta: float,
               maximizing_player: bool) -> float:
        return board.ghost_test(from_pos, to_pos, 
                                lambda x: self.minimax_wrapper(x, depth-1, not maximizing_player, alpha, beta))

    def minimax_wrapper(self,
                       board: Board,
                       depth: int,
                       maximizing_player: bool,
                       alpha: float,
                       beta: float) -> float:
        
        if self.debug:
            debcolor = Color.RED if maximizing_player else Color.BLACK

        if depth == 1:
            evaluate = board.evaluate(self.eval_set)
            if self.debug:
                print(f"ENGINE WRAPPER: Evaluation = {evaluate}")
                # board.print_visual()
            return evaluate
        
        extreme_value = -math.inf if maximizing_player else math.inf
        pieces = board.get_reds() if maximizing_player else board.get_blacks()

        for piece in pieces:
            valid_moves = board.get_piece_valid_moves(piece, check=False)
            for move in valid_moves:
                if self.debug:
                    print("--------------------------------------" * depth)
                    print(f"MINIMAX {debcolor} with {piece.get_position()}->{move}")
                    print(f"            depth={depth}, alpha={alpha}, beta={beta}")
                current_value = self.minimax(
                    board=board,
                    from_pos=piece.get_position(),
                    to_pos=move,
                    depth=depth,
                    alpha=alpha,
                    beta=beta,
                    maximizing_player=maximizing_player)
                
                if self.debug:
                    print(f"MINIMAX EVALUATION for {debcolor} after {piece.get_position()}->{move}: {debcolor.opposite().name} can get {current_value}")
                    print("--------------------------------------" * depth)

                if maximizing_player:
                    extreme_value = max(extreme_value, current_value)
                    alpha = max(alpha, extreme_value)
                else:
                    extreme_value = min(extreme_value, current_value)
                    beta = min(beta, extreme_value)

                if beta <= alpha:
                    break

            if beta <= alpha:
                break

        return extreme_value

File: game_logic/new_clang/clang/engine.cpp
#include "engine.h"

Engine::Engine(const EvaluateSet& eval_set, int depth, bool debug)
    : eval_set(eval_set), depth(depth), debug(debug) {
    std::srand(static_cast<unsigned>(std::time(nullptr)));
}

int Engine::depth_control(int num_children) {
    if (num_children <= 20) {
        return 4;
    } else if (num_children <= 30){
        return 2;
    }
    return 0;
}

std::pair<double, std::optional<std::pair<Vector, Vector>>> Engine::get_best_move(Board& board, Color current_player, bool is_random) {
    double alpha = -std::numeric_limits<double>::infinity();
    double beta = std::numeric_limits<double>::infinity();
    std::vector<std::pair<Vector, Vector>> best_moves;
    std::optional<std::pair<Vector, Vector>> best_move = std::nullopt;

    const auto& pieces = (current_player == Color::RED) ? board.get_reds() : board.get_blacks();
    std::vector<std::tuple<double, Vector, Vector>> moves;

    for (const auto& piece : pieces) {
        Vector from_pos = piece->get_position();
        std::vector<Vector> valid_moves = board.get_piece_valid_moves(piece.get(), true);

        for (const Vector& to_pos : valid_moves) {
            double eval_bonus = 0.0;
            auto target_square = board.get_square(to_pos);
            auto captured_piece = target_square.get_piece();

            if (captured_piece) {
                eval_bonus += captured_piece->get_value() * 0.5;
            }

            if (piece->is_attacker()) {
                eval_bonus += 0.3;
            }

            moves.emplace_back(eval_bonus, from_pos, to_pos);
        }
    }

    if (current_player == Color::RED) {
        std::sort(moves.begin(), moves.end(), [](const auto& a, const auto& b) {
            return std::get<0>(a) > std::get<0>(b);
        });
    } else {
        std::sort(moves.begin(), moves.end(), [](const auto& a, const auto& b) {
            return std::get<0>(a) < std::get<0>(b);
        });
    }

    if (!moves.empty()) {
        best_move = std::make_pair(std::get<1>(moves[0]), std::get<2>(moves[0]));
    }

    int depth_boost = 0;
    //int depth_boost = depth_control(moves.size());
    //std::cout << "DEPTH BOOST: " << depth_boost << std::endl;

    for (const auto& move_eval : moves) {
        double bonus = std::get<0>(move_eval);
        Vector from_pos = std::get<1>(move_eval);
        Vector to_pos = std::get<2>(move_eval);

        if (debug) {
            std::cout << "\n==================================================================\n";
            std::cout << "ENGINE for " << (current_player == Color::RED ? "RED" : "BLACK") 
                      << " with " << from_pos << "->" << to_pos << ":\n";
        }

        double move_value = minimax(
            board, from_pos, to_pos, depth + depth_boost, alpha, beta, (current_player == Color::RED)
        );

        if (debug) {
            std::cout << "ENGINE EVALUATION: " << (current_player == Color::RED ? "BLACK" : "RED") 
                      << " can get " << move_value << "\n";
            std::cout << "==================================================================\n\n";
        }

        if (current_player == Color::RED) {
            if (move_value > alpha) {
                if (is_random && (alpha == -std::numeric_limits<double>::infinity() || move_value / alpha < 1.05)) {
                    best_moves.emplace_back(from_pos, to_pos);
                } else {
                    best_move = std::make_pair(from_pos, to_pos);
                }
                if (move_value == std::numeric_limits<double>::infinity()) break;
                alpha = move_value;
            }
        } else {
            if (move_value < beta) {
                if (is_random && (beta == std::numeric_limits<double>::infinity() || beta / move_value < 1.05)) {
                    best_moves.emplace_back(from_pos, to_pos);
                } else {
                    best_move = std::make_pair(from_pos, to_pos);
                }
                if (move_value == -std::numeric_limits<double>::infinity()) break;
                beta = move_value;
            }
        }
    }

    double move_value = (current_player == Color::RED) ? alpha : beta;
    if (is_random) {
        if (!best_moves.empty()) {
            int index = std::rand() % best_moves.size();
            return { move_value, best_moves[index] };
        }
        return { move_value, handle_gameover(board, current_player) };
    }
    return best_move ? std::make_pair(move_value, best_move) 
                     : std::make_pair(move_value, handle_gameover(board, current_player));
}

std::optional<std::pair<Vector, Vector>> Engine::handle_gameover(Board& board, Color color) {
    // bool old_debug = board.is_debug();
    // board.is_debug() = true;

    const auto& team = (color == Color::RED) ? board.get_reds() : board.get_blacks();
    for (const auto& piece : team) {
        std::vector<Vector> vds = board.get_piece_valid_moves(piece.get(), true);
        //std::cout << "ENGINE GAMEOVER: " << piece->get_name() << " valid moves: ";
        //for (const auto& v : vds) std::cout << v << " ";
        //std::cout << "\n";
    }

    // board.is_debug() = old_debug;
    return std::nullopt;
}

double Engine::minimax(Board& board, const Vector& from_pos, const Vector& to_pos, int depth, double alpha, double beta, bool maximizing_player) {
    return board.ghost_test(from_pos, to_pos, [&](Board& b) {
        return minimax_wrapper(b, depth - 1, !maximizing_player, alpha, beta);
    });
}

double Engine::minimax_wrapper(Board& board, int depth, bool maximizing_player, double alpha, double beta) {
    if (depth == 0) {
        double eval = board.evaluate(eval_set);
        if (debug) std::cout << "ENGINE WRAPPER: Evaluation = " << eval << "\n";
        return eval;
    }

    double extreme_value = maximizing_player 
        ? -std::numeric_limits<double>::infinity() 
        : std::numeric_limits<double>::infinity();
    
    const auto& pieces = maximizing_player ? board.get_reds() : board.get_blacks();

    for (const auto& piece : pieces) {
        Vector piece_pos = piece->get_position();
        std::vector<Vector> valid_moves = board.get_piece_valid_moves(piece.get(), false, false);

        for (const Vector& move : valid_moves) {
            if (debug) {
                std::cout << "--------------------------------------\n";
                std::cout << "MINIMAX " << (maximizing_player ? "RED" : "BLACK")
                          << " with " << piece_pos << "->" << move << "\n";
                std::cout << "depth=" << depth << ", alpha=" << alpha << ", beta=" << beta << "\n";
            }

            double current_value = minimax(
                board, piece_pos, move, depth, alpha, beta, maximizing_player
            );

            if (debug) {
                std::cout << "MINIMAX EVALUATION: " << (maximizing_player ? "BLACK" : "RED") 
                          << " can get " << current_value << "\n";
                std::cout << "--------------------------------------\n";
            }

            if (maximizing_player) {
                extreme_value = std::max(extreme_value, current_value);
                alpha = std::max(alpha, extreme_value);
            } else {
                extreme_value = std::min(extreme_value, current_value);
                beta = std::min(beta, extreme_value);
            }

            if (beta <= alpha) break;
        }
        if (beta <= alpha) break;
    }

    return extreme_value;
}